hash_practice/exercises.py

```diff
@@ -2,28 +2,224 @@
+            hash_table[hash_index] = [string]
+        else:
+            hash_table[hash_index].append(string)
+
+    return list(hash_table.values())
+
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+
+        Pseudocode solution 2:
+        1.Go over the input array and create unique element frequency hash table.
+        2. Go over unique element frequency hash table and create new hash table with key being frequencies and values being list of unique element that share that same frequency.
+        3. While iterating over unique element frequency hash table calculate the max availible frequency.
+        4. Iterate over key-value pairs in descending order (starting from max frequency) and construct resulting array by adding k elements to it.
+
+        This solution would work for both sorted and not sorted input array.
+
+        Time Complexity: O(n)
+        Space Complexity: O(n)
+    """
+    if len(nums) == k:
+        return nums
+    if len(nums) == 0:
+        return []
+    frequency_cells = {}
+    for num in nums:
+        if frequency_cells.get(num):
+            frequency_cells[num] += 1
+        else:
+            frequency_cells[num] = 1
+
+    max_frequency = 0
+    key_frequency_cells = {}
+    for num, frequency in frequency_cells.items():
+        if not key_frequency_cells.get(frequency):
+            key_frequency_cells[frequency] = [num]
+        else:
+            key_frequency_cells[frequency].append(num)
+        if frequency > max_frequency:
+            max_frequency = frequency
+
+    result = []
+    while max_frequency > 0 and len(result) < k:
+        if key_frequency_cells.get(max_frequency):
+            for num in key_frequency_cells[max_frequency]:
+                if len(result) < k:
+                    result.append(num)
+        max_frequency -= 1
+
+    return result
+
+
+def top_k_frequent_elements_alternative_solution(nums, k):
+    """ This method will return the k most common elements
+        In the case of a tie it will select the first occuring element.
+
+        pseudocode solution 1:
+        This solution would work for both sorted and not sorted input array.
+        0. Go over the input array and create unique element frequency hash table.
+        1. Go over the frequency hash table and create an array of tuples: [(x1,y1),(x2,y2),(x3,y3),...(xn,yn)]
+        where x1, x2, ..., xn - unique integer elements from the input array
+        and y1, y2, ..., yn their frequencies relatively.
+        2. Sort new array by key, where key is frequency (second element of each tuple).
+        3. Iterate over sorted array k times to form the output by grabbing first element of each tuple and appending it to the resulting array
+
+        Time Complexity: O(nlogn) (sorting)
+        Space Complexity: O(n+k) ?
+
     """

```

Comment:
👍 This is the other standard solution, nice.

Either O(n + k) or O(n) are fine for space complexity since k < n